pkg/kubelet/kubelet_node_status.go

```diff
@@ -39,67 +39,117 @@ import (
-			}
-			glog.Errorf(
-				"Previously %q had externalID %q; now it is %q; will delete and recreate.",
-				kl.nodeName, node.Spec.ExternalID, currentNode.Spec.ExternalID,
-			)
-			if err := kl.kubeClient.Core().Nodes().Delete(node.Name, nil); err != nil {
-				glog.Errorf("Unable to delete old node: %v", err)
+		glog.Infof("Attempting to register node %s", node.Name)
+		registered := kl.registerWithApiServerAttempt(node)
+		if registered {
+			glog.Infof("Successfully registered node %s", node.Name)
+			kl.registrationCompleted = true
+			return
+		}
+	}
+}
+
+// registerWithApiServerAttempt makes an attempt to register the given node
+// with the API server, returning a boolean indicating whether the attempt was
+// successful.  If a node with the same name already exists, it reconciles the
+// value of the annotation for controller-managed attach-detach of attachable
+// persistent volumes for the node.  If a node of the same name exists but has
+// a different externalID value, it attempts to delete that node so that a
+// later attempt can recreate it.
+func (kl *Kubelet) registerWithApiServerAttempt(node *api.Node) bool {
+	_, err := kl.kubeClient.Core().Nodes().Create(node)
+	if err == nil {
+		return true
+	}
+
+	if !apierrors.IsAlreadyExists(err) {
+		glog.Errorf("Unable to register node %q with API server: %v", kl.nodeName, err)
+		return false
+	}
+
+	existingNode, err := kl.kubeClient.Core().Nodes().Get(kl.nodeName)
+	if err != nil {
+		glog.Errorf("Unable to register node %q with API server: error getting existing node: %v", kl.nodeName, err)
+		return false
+	}
+	if existingNode == nil {
+		glog.Errorf("Unable to register node %q with API server: no node instance returned", kl.nodeName)
+		return false
+	}
+
+	if existingNode.Spec.ExternalID == node.Spec.ExternalID {
+		glog.Infof("Node %s was previously registered", kl.nodeName)
+
+		// Edge case: the node was previously registered; reconcile
+		// the value of the controller-managed attach-detach
+		// annotation.
+		var (
+			existingCMAAnnotation, _ = existingNode.Annotations[volumehelper.ControllerManagedAttachAnnotation]
+			newCMAAnnotation, newSet = node.Annotations[volumehelper.ControllerManagedAttachAnnotation]
+		)
+
+		if newCMAAnnotation != existingCMAAnnotation {
+			// If the just-constructed node and the current node do
+			// not have the same value, update the current node with
+			// the correct value of the annotation.
+
+			if !newSet {
+				glog.Info("Controller attach-detach setting changed to false; updating existing Node")
+				delete(existingNode.Annotations, volumehelper.ControllerManagedAttachAnnotation)
 			} else {
-				glog.Errorf("Deleted old node object %q", kl.nodeName)
+				glog.Info("Controller attach-detach setting changed to true; updating existing Node")
+				existingNode.Annotations[volumehelper.ControllerManagedAttachAnnotation] = newCMAAnnotation
+			}
+
+			if _, err := kl.kubeClient.Core().Nodes().Update(existingNode); err != nil {
```

Comment:
@ncdc I think that would be equally applicable to any resource Kubelet can update, wouldn't it?